WIP: update EtcdBackendQuota feature (#2031)
openshift/enhancements: enhancements/etcd/etcd-size-tuning.md

Reviewer: This still says `N/A`. We should probably explain why before going GA.

Can we capture https://github.com/openshift/enhancements/pull/2031#discussion_r3374038831 here.

Perhaps something like:

```suggestion
No special handling is required. The `backendQuotaGiB` field and its value are preserved across upgrades and z-stream rollbacks. Y-stream downgrades are not supported by the CVO. If a z-stream rollback occurs, the operator continues to respect the configured quota value.
```

Or anything else that I missed.

Author: Thx for the suggestion.

api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
  - None
creation-date: 2024-01-24

> last-updated: 2026-06-11

Reviewer: Nit: this should match your commit date I think.
`2026-07-23` currently. Or whatever the next one is if you do update.

* Allow configuration of the backend limit via human readable units: 16GiB.
* Add an API to allow admins to change the value.

> * The backend cannot be decreased below the current DB size.

Reviewer 2: In the general context of allowing decreases. I'm wondering if that should be a separate, future enhancement, and that we should just focus on getting the GA of this feature as-is.

Like with the lower max limitation, it's allowable to relax constraints in z-streams which I would argue this would be.

I think, in general, the customers asking for this feature are doing so because they're running against the current quota and wish to increase it forever. Given the memory consumption implications of increasing the quota (and using the headroom gained by increasing the quota), ensuring the sizing of the nodes prior to increase is essential, and I think that many, if not all, customers will view this as a one-time option. Because of those implications and requirements, the node will be sized larger anyway, regardless of what the quota is. Meaning, if they provision nodes that can handle 16GiB size-on-disk (and set the quota to 16GiB), and they're only using 12GiB on-disk. They would then need to re-provision the node to be smaller: just changing the quota doesn't directly effect how much memory or storage etcd takes, it's just a limit.

That is all to say, I don't know if allowing decrease is a feature that will be used, and in my opinion, it would be better to wait for a customer to ask for it - so we know there's a usecase - before putting in the work to implement it for this release cycle.

Author: I still think, downsizing your cluster in order to reduce cost is still relevant part of the story, but it can be done later. I agree that the feature can be pushed as is, to not postpone the graduation. I moved the Quota decrease story to the Alternatives.

We will need to work with the perf team to understand the memory requirements of largers quotas, and the effect of the quota on things like defragmentation, snapshot restore, etc.
A downside is that with the profiles, discussed in the Alternatives section, there are a discrete, small, number of permutations to test, giving more confidence of the exact effects a given profile has on performance/memory requirements.
This is a minor consideration as setting the quota to a arbitrarily higher value does not mean that etcd will take up that much space; it will depend on the workload.

### Decreasing the Quota

Decreasing the quota has the following benefits:
- The DB size has a big impact on the memory usage. Decreasing the quota will limit the memory usage.
  This can prevent further cluster performance degradation

> - During a cluster revert to a former state (downgrade, backup/restore) it might be beneficial to keep the old value. This also affects alerts / expectations and other metrics. 

Reviewer 2: Downgrades should not be a factor as the [CVO does not allow y-stream downgrades](https://github.com/openshift/cluster-version-operator/blob/4cfd94b468732a9fff4c29afa679085fcf363e5f/pkg/payload/precondition/clusterversion/rollback.go#L71-L1100) (since 4.14 or so); only z-stream from the one version previous to an upgrade.

Backup and restore should also not be a factor since restoring effectively moves the cluster back in time. Since this is a wholly destructive process, it does not matter what the current quota or current size is, they'll be reverted to the previous quota and all objects in etcd will be replaced with the previous state.

Author: It might also include partial backup/restore done by other tools. Nevertheless, I removed it as it is not the most important flow. I am only highlighting the memory impact now.

#### Invalid value

> 1. The cluster administrator wants to set the quota from a larger value (e.g. 32GiB) to a smaller one (8GiB). The current size of the DB is 10 Gib and is reported in the `.status.BackendSizeKiB`.

Reviewer 2: Replace 32GiB with our new limit of 16GiB

2. They attempt to set the new value via the `etcd.operator.openshift.io/v1` API.
3. The API server should return an error notifying them that the value is invalid - it is not allowed to shrink the quota below the current DB size.

### API Extensions

In operator/v1/0000_12_etcd-operator_01_config-TechPreviewNoUpgrade.crd.yaml:
- Add BackendQuotaGiB of type int32
  - With default '8GiB'.
  - With validation disallow a decrease in value below the BackendSizeKiB (with a buffer).

>   - With an upper limit of '32GiB'.

Reviewer 2: 16GiB

## Open Questions [optional]

For each of these, we would need to compare against the current default of 8GiB. Increasing the quota does not necessarily increase the following, but at the extreme case (e.g. increase to 16GiB and using nearly that much) we expect the following to perform differently.

* How long does it take to compact & defrag a database of a given size?

>   * Defragmentation will pause writes to etcd, reads from etcd will be unaffected.

Reviewer: Not a blocker but for a GA feature where larger DBs will make defrag
take longer this seems like it would be useful to somewhat answer this.

Given that we previously held this to asses defrag timing impact (https://github.com/openshift/enhancements/pull/2031#issuecomment-4993028601) did we ever capture that anywhere?
If so it might be good to list that out in the Risks section similar to the snapshot sync timings.

Or closing the question with a note if we've determined that this is workload dependent enough that the numbers aren't too meaningful.
And that the automated defrag in the etcd-operator makes it less of an admin facing concern.

Author: I moved this from the open questions to the risks section.

@dusk125 is trying to improve on the defrag, but it is not blocking this feature.

Author: Added measurements and risks for the defrag.

## Graduation Criteria

> None

Reviewer: Since we have it defined now
```suggestion
```

* Larger:
  - ETCD_BACKEND_QUOTA: 16GiB

This alternative was rejected because the discrete values don't provide a benefit to performance or safety of the etcd cluster: it just allows the etcd database to be larger if it needs to be. Also, the current enhancement proposal allows for users to more closely configure their etcd cluster to the resources available to them; with discrete values, a user could have to choose between a value that's too small or too large for their purposes.

### Decreasing the Quota

In general increasing the quota is done in response to increased demand on the cluster.
Decreasing or reverting the quota is unlikely to happen often, but it has the following benefit:

> - The DB size has a big impact on the memory usage. Decreasing the quota will limit the memory usage.

Reviewer 2: > Decreasing the quota will limit the memory usage

That's slightly misleading. Decreasing the quota would allow a customer to downsize a node (to reduce cost) which I think we should highlight as the main motivation of this enhancement; just changing the quota only affects the etcd instances if they're above/at the quota: then they can't write, but if they're below the large quota, their memory usage is no different than if the quota was smaller.

Author: +1, I updated this part. Changing the quota has to used correctly together with monitoring IMO, to get the main benefit. I am keeping this there, but the downsizing is also important. Please let me know if I should highlight it further.

Reviewer 2: I think your update is clearer!

## Support Procedures

- If the user attempts to set the quota to an invalid value (not in the form of a resource quantity (e.g. 8GiB, etc)), the API will not accept the value and return an error.

> - If the user attempts to set the quota to a lower value than the current DB size (e.g. 21GiB -> 16GiB), the API will not accept the value and return an error.

Reviewer: Looking at the validation today:
> // +kubebuilder:validation:XValidation:rule="self>=oldSelf",message="etcd backendQuotaGiB may not be decreased"

We're not allowing any decrease period, not just lower than the current DB size right?
This wording makes it seem like we can decrease the quota, just not lower than the current DB size.

If I'm correct then perhaps something more clear like:

```suggestion
- If the user attempts to decrease the quota (e.g. 16GiB -> 8GiB), the API will not accept the value and return an error — the backend quota cannot be lowered once set.
```

Author: +1, this is more precise

appropriate. To implement the new validation, we would also report the current DB size in
`.status.BackendSizeKiB`.

```go

// +kubebuilder:validation:XValidation:rule="self.spec.backendQuotaGiB >= oldSelf.spec.backendQuotaGiB || (self.spec.backendQuotaGiB * 1024 * 1024) >= (self.status.BackendSizeKiB + 100 * 1024)",message="etcd .spec.backendQuotaGiB must be greater than .status.backendSizeKiB plus a 100 MiB buffer"
type Etcd struct {
  Spec EtcdSpec `json:"spec"`
  Status EtcdStatus `json:"status"`
}

type EtcdSpec struct {
	// backendQuotaGiB sets the etcd backend storage size limit in gibibytes.
	// The value should be an integer not less than 8 and not more than 16.
	// When not specified, the default value is 8.
	// +kubebuilder:default:=8
	// +kubebuilder:validation:Minimum=8
	// +kubebuilder:validation:Maximum=16
	// +openshift:enable:FeatureGate=EtcdBackendQuota
	// +default=8
	// +optional
  BackendQuotaGiB int32 `json:"backendQuotaGiB"`
}

type EtcdStatus struct {
  // BackendSizeKiB represents the current etcd backend storage database size in kibibytes.
  // +kubebuilder:validation:Minimum=0
  // +openshift:enable:FeatureGate=EtcdBackendQuota
  // +optional

>   BackendSizeKiB *int64 `json:"backendSizeMiB"`

Reviewer: Nit, if you meant KiB here:
```suggestion
  BackendSizeKiB *int64 `json:"backendSizeKiB"`
```

Reviewer 3: should this be a resource.Quantity instead so that it's a bit more human readable?

Edit: typo, accidentally wrote resource.Quota at first

Author: Do you mean introducing a new API, or a new field? Nevertheless, this is an Alternative that we might do in the future. We can review the API again if we do that.

Reviewer 3: I meant replacing the field with:
```
BackendSize *resource.Quantity `json:"backendSize"`
```

so that it gets serialized in binary SI units

Reviewer 3: But yeah not important for this as it's a future enhancement

Author: That would be more user friendly, but I am not sure if it would work well with the CEL validation. See `XValidation` above Etcd struct.

I will keep it as is for now, and let's potentially revisit.

Reviewer 3: True, it does not support numeric validations unfortunately
